### PR TITLE
[IE CLDNN] Convolution fsv16 occupancy improvements with SLM using

### PR DESCRIPTION
**Pull request by lznamens**
These changes apply to block convolution improvements for new GPUs that have separate SLM.

JIRA: 42409, 45430

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16_1x1.cl [hunk 1]
```diff
@@ -37,16 +45,21 @@ KERNEL(convolution_b_fs_yx_fsv16_1x1)(
...
-    const int y = (xy * X_BLOCK_SIZE) / OUTPUT_SIZE_X;
+    const uint input_x = x;
+    const uint input_y = y;
+#endif
+    const uint b = (int)get_global_id(2);
+    const uint lid = (int)get_sub_group_local_id();
 
-    const int input_x = x;
-    const int input_y = y;
+    const uint lid1 = (int)get_local_id(1);
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16_1x1.cl:59]: Same as for generic kernel

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16_1x1.cl [hunk 2]
```diff
@@ -83,71 +96,92 @@ KERNEL(convolution_b_fs_yx_fsv16_1x1)(
...
-    vec_t dst = (vec_t)(UNIT_BLOCK_READ(biases, f_block * FEATURE_SLICE_SIZE));
+#if SLM_DIV_FACTOR == 1
+    vec_t dst = (vec_t)(UNIT_BLOCK_READ(biases, feature_block * FEATURE_SLICE_SIZE));
+#else
+    vec_t dst;
+
+    if (feature_sub_block == 0) {
+        dst = (vec_t)(UNIT_BLOCK_READ(biases, feature_block * FEATURE_SLICE_SIZE));
+    }
+    else {
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16_1x1.cl:116]: Same as for generic kernel

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16_1x1.cpp [hunk 3]
```diff
@@ -137,31 +185,42 @@ JitConstants ConvolutionKernel_b_fs_yx_fsv16_1x1::GetJitConstants(const convolut
...
     auto blockWidth = dispatchData.cldnnStyle.blockWidth;
     if (!params.fused_ops.empty()) {
         auto input_dt = GetUnitType(params);
         FusedOpsConfiguration conf_vec = { "_VEC",
-                                           {"b", "(f_block*16)", "y", "x"},
+                                           {"b", "(feature_block * 16)", "y", "x"},
                                            "dst",
                                            input_dt,
-                                           blockWidth,
+                                           blockWidth > 1 ? blockWidth : 2,
```
Comment (sshlyapn) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16_1x1.cpp:197]: Why do you need this hack?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16.cpp [hunk 4]
```diff
@@ -55,6 +52,48 @@ ConvolutionKernel_b_fs_yx_fsv16::AutoTuneOption ConvolutionKernel_b_fs_yx_fsv16:
...
+                                                         const ConvolutionTuningData& tuning_data) const {
+    auto tuneOptions = GetAutoTuneOptions(params, 0);
+    auto blockWidth = tuneOptions.blockWidth;
+
+    auto x = params.output.X().v;
+    auto y = params.output.Y().v;
+    auto f = params.output.Feature().v;
+    auto b = params.output.Batch().v;
+
+    auto threads = CeilDiv(x * y, blockWidth) * CeilDiv(f, tuning_data.feature_block_size) * tuning_data.slm_div_factor * b;
```
[the remark this answers is not in the document]
Comment (lznamens) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16.cpp]: Thanks. These are results of copy-past. Fixed.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16.cpp [hunk 5]
```diff
@@ -55,6 +52,48 @@ ConvolutionKernel_b_fs_yx_fsv16::AutoTuneOption ConvolutionKernel_b_fs_yx_fsv16:
...
+}
+
+ConvolutionKernel_b_fs_yx_fsv16::ConvolutionTuningData ConvolutionKernel_b_fs_yx_fsv16::GetTuningParams(const convolution_params& params) const {
+    ConvolutionTuningData tuning_data;
+
+    const auto& input = params.inputs[0];
+
+    size_t feature_blocks_count = input.Feature().v % tuning_data.feature_block_size != 0 ?
+                                  input.Feature().v / tuning_data.feature_block_size :
+                                  CeilDiv(input.Feature().v, tuning_data.feature_block_size);
```
[the remark this answers is not in the document]
Comment (lznamens) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16.cpp]: Fixed.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16_1x1.cpp [hunk 6]
```diff
@@ -50,6 +53,46 @@ ConvolutionKernel_b_fs_yx_fsv16_1x1::AutoTuneOption ConvolutionKernel_b_fs_yx_fs
...
+}
+
+ConvolutionKernel_b_fs_yx_fsv16_1x1::ConvolutionTuningData ConvolutionKernel_b_fs_yx_fsv16_1x1::GetTuningParams(const convolution_params& params) const {
+    ConvolutionTuningData tuning_data;
+
+    const auto& input = params.inputs[0];
+
+    size_t feature_blocks_count = input.Feature().v % tuning_data.feature_block_size != 0 ?
+                                  input.Feature().v / tuning_data.feature_block_size :
+                                  CeilDiv(input.Feature().v, tuning_data.feature_block_size);
```
[the remark this answers is not in the document]
Comment (lznamens) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_b_fs_yx_fsv16_1x1.cpp]: Fixed.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16.cl [hunk 7]
```diff
@@ -114,6 +114,11 @@ KERNEL(convolution_bfyx_f16)(
     const int x = (xy % X_BLOCKS) * OUTPUT_X_BLOCK_SIZE;
     const int y = (xy / X_BLOCKS);
 
+    const int lid1 = (int)get_local_id(1);
```
[the remark this answers is not in the document]
Comment (lznamens) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16.cl:106]: Fixed.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16.cl [hunk 8]
```diff
@@ -95,14 +97,12 @@ KERNEL(convolution_bfyx_f16)(
 #endif
     uint split_idx) {
 #if GROUPED
-    const int f_block = get_group_id(1);
-    const int group = (f_block * FEATURE_SLICE_SIZE) / FILTER_OFM_NUM;
-    const int prev_group_leftover = (FILTER_OFM_NUM * (group + 1)) - (f_block * FEATURE_SLICE_SIZE);
+    const int group = (feature_block * FEATURE_SLICE_SIZE) / FILTER_OFM_NUM;
```
[the remark this answers is not in the document]
Comment (lznamens) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_f16.cl]: Yes, thanks. Fixed.

**Review by sshlyapn — APPROVED** [review]

**Pull request merged**